base/reflection.jl

```diff
@@ -348,17 +346,41 @@ function raise_match_failure(name::Symbol, @nospecialize(tt))
     error("$name: unanalyzable call given $sig_str")
 end
 
+"""
+    default_compiler_module()
+
+In order to facilitate working on the compiler, we'd like to make it easy for
+our users to switch all the ordinary compiler-based reflection to a new copy of the
+compiler, if they `dev` and load a new version of the compiler. However, it would
+be confusing if the result of reflection functions changed, simply because some
+downstream package happens to load a new copy of `Compiler`. Thus, we look for
+a special rendezvous binding `var"#_interactive_compiler_reference` in Main.
+This binding is exported by the compiler, so if the user does `using Compiler` to
+load a dev'ed version of the Compiler from `Main`, we will see that and pick up
+the new Compiler, but if the using statement is in some downstream package, we
+will not.
+"""
+function default_compiler_module()
+    # If the user has explicitly loaded a new compiler into Main, prefer that.
+    if isdefined(Core.Main, :var"#_interactive_compiler_reference")
+        return Core.Main.var"#_interactive_compiler_reference"
+    end
+    # Otherwise return a private reference to the internal compiler
+    return Base.Compiler
+end
+
 """
     code_typed_by_type(types::Type{<:Tuple}; ...)
 
 Similar to [`code_typed`](@ref), except the argument is a tuple type describing
 a full signature to query.
 """
 function code_typed_by_type(@nospecialize(tt::Type);
+                            Compiler::Module = default_compiler_module(),
```

Comment:
This will make all this code non-statically-inferable. Are the juliac folks okay with this?

Reply:
I figured it was already pretty dynamic and interactive. It if becomes a problem, we can lift it to the type level to specialize.

Reply:
This kind of reflection in `juliac` would require getting all of inference, etc. juliac-compatible, right?

We're still a ways away from that so I agree, probably fine.

Reply:
> This kind of reflection in `juliac` would require getting all of inference, etc. juliac-compatible, right?

Yes